File: apps/companion/native/ios/KalaReachNative/VoiceCall.swift
//
//  The native voice call: WebRTC's own peer connection, and the microphone the platform owns.
//
//  Section 15 ¶2 is the constraint this file exists to satisfy, and it admits no exception: native
//  WebRTC and native platform audio own capture and playback, not a background WebView
//  `getUserMedia` path. Nothing here goes near the WebView. The offer is made by this process, the
//  answer is applied by this process, and the audio flows between this device and the provider
//  without passing through the interface or through KalaReach.
//
//  Section 15 ¶6 is the other constraint: the provider's data channel is read-only. This
//  connection creates no data channel of its own and sends **zero** bytes on the one the provider
//  opens. What arrives is handed to the caller as bytes with a type, and an event the caller does
//  not recognise is dropped rather than reflected into anything.
//

import Foundation
import WebRTC

/// What a running call tells the application about.
public protocol VoiceCallObserver: AnyObject {
    /// The connection's own state changed.
    func voiceCall(_ call: VoiceCall, connectionChanged state: RTCPeerConnectionState)
    /// The provider sent something on its read-only channel.
    func voiceCall(_ call: VoiceCall, receivedProviderEvent data: Data)
    /// The first remote audio arrived. Section 27's KR-PERF-010 first-audio figure is taken here.
    func voiceCallReceivedFirstAudio(_ call: VoiceCall)
}

/// One native voice call.
///
/// It owns exactly three things: the peer connection, the local microphone track, and the remote
/// audio track the provider sends. Everything about *who* is on the other end — the broker, the
/// account, the control socket — is outside it, which is what keeps the provider interface modular
/// in §15 ¶1's sense: a different provider replaces the signalling around this object and replaces
/// nothing inside it.
///
/// A call has two stages, and the microphone belongs only to the second. Building it makes the
/// connection and a microphone track that is off, for the offer and the answer; no audio session
/// is opened and nothing is captured. ``permit(voiceSessionId:closesAtEpochMs:)`` is given the
/// host's answer to the start, and only then is the session opened and capture allowed. Whether
/// the microphone is on is decided by one ``VoiceCaptureGate`` from the permit, the person's mute,
/// what the system did and the route, and the call stops itself when its deadline comes.
public final class VoiceCall: NSObject {
    /// Shared across calls, because building one is expensive and it holds the audio device.
    private static let factory: RTCPeerConnectionFactory = {
        RTCInitializeSSL()
        return RTCPeerConnectionFactory(
            encoderFactory: RTCDefaultVideoEncoderFactory(),
            decoderFactory: RTCDefaultVideoDecoderFactory()
        )
    }()

    private let connection: RTCPeerConnection
    private let microphone: RTCAudioTrack
    private weak var observer: VoiceCallObserver?
    private var announcedFirstAudio = false

    /// Held by every change to what the microphone and the speaker are allowed to do.
    ///
    /// Recursive, because opening or closing the audio session can report a route change on the
    /// thread that is doing it, and that report takes this lock too.
    private let lock = NSRecursiveLock()
    private let gate = VoiceCaptureGate()
    private var stopped = false
    private var expiry: DispatchWorkItem?

    /// Whether the person has muted their own microphone.
    public var isMutedByPerson: Bool { gate.isMutedByPerson }

    /// Whether the remote voice is being played out of this device.
    public private(set) var isPlaybackMuted = false

    /// Builds a call for its offer and its answer, with the microphone off.
    ///
    /// No audio session is opened here: that waits for ``permit(voiceSessionId:closesAtEpochMs:)``,
    /// so nothing but a call the host permitted can open the microphone.
    public init(observer: VoiceCallObserver) throws {
        let configuration = RTCConfiguration()
        // The provider's answer names its own candidates. No KalaReach STUN or TURN server is
        // configured here: media travels between this device and the provider, and a relay of
        // KalaReach's would be a third party in a path §15 ¶3 says has two.
        configuration.sdpSemantics = .unifiedPlan
        configuration.continualGatheringPolicy = .gatherContinually

        let constraints = RTCMediaConstraints(
            mandatoryConstraints: nil,
            // The one place the caller could have asked for a data channel of its own. It does
            // not: §15 ¶6 makes the provider's channel read-only and this end creates none.
            optionalConstraints: nil
        )
        guard
            let made = VoiceCall.factory.peerConnection(
                with: configuration,
                constraints: constraints,
                delegate: nil
            )
        else {
            throw VoiceAudioError.answerNotApplicable("this device could not open a connection")
        }
        connection = made

        // Audio processing is the platform's. On iOS libwebrtc's audio device runs through Apple's
        // Voice-Processing I/O unit, which is where echo cancellation, gain control and noise
        // suppression happen; asking for them again in software would be two cancellers fighting.
        let source = VoiceCall.factory.audioSource(with: RTCMediaConstraints(
            mandatoryConstraints: nil,
            optionalConstraints: nil
        ))
        microphone = VoiceCall.factory.audioTrack(with: source, trackId: "kr-voice-microphone")
        // Off until the call is permitted. The offer describes a track, and a described track
        // carries nothing until the gate lets it.
        microphone.isEnabled = false
        self.observer = observer
        super.init()
        connection.delegate = self
        connection.add(microphone, streamIds: ["kr-voice"])
    }

    /// This device's monotonic clock, in milliseconds. The gate's deadlines are on it.
    private static func nowMs() -> UInt64 { DispatchTime.now().uptimeNanoseconds / 1_000_000 }

    /// The furthest deadline a call is given, in milliseconds: a day.
    private static let longestCallMs: UInt64 = 86_400_000

    /// Opens the microphone for a call the host started.
    ///
    /// Takes the host's answer: the voice session and the moment the service closes the call, in
    /// UTC milliseconds. The audio session is opened here and nowhere else, and the call stops
    /// itself when that moment comes, whether or not anything else happened.
    ///
    /// - Returns: false, and nothing opened, when this call is stopped, already permitted or past
    ///   its deadline.
    /// - Throws: ``VoiceAudioError/microphoneNotPermitted`` when the microphone is not available.
    public func permit(voiceSessionId: String, closesAtEpochMs: UInt64) throws -> Bool {
        lock.lock()
        defer { lock.unlock() }
        guard !stopped, gate.current == nil else { return false }
        let wall = UInt64(Date().timeIntervalSince1970 * 1_000)
        guard closesAtEpochMs > wall else { return false }
        let now = VoiceCall.nowMs()
        // No call runs for a day. A deadline further away than that closes the call at a day, never
        // later, and keeps the arithmetic below inside its type whatever the answer said.
        let deadline = now + min(closesAtEpochMs - wall, VoiceCall.longestCallMs)
        try AudioSession.shared.activate(for: self)
        guard gate.permit(voiceSessionId: voiceSessionId, deadlineMs: deadline, nowMs: now) != nil else {
            try? AudioSession.shared.deactivate()
            return false
        }
        let ends = DispatchWorkItem { [weak self] in self?.stop() }
        expiry = ends
        DispatchQueue.main.asyncAfter(
            deadline: .now() + .milliseconds(Int(deadline - now)),
            execute: ends
        )
        apply(now)
        return true
    }

    /// Whether the microphone was carrying speech at `atMs` on this device's monotonic clock.
    public func couldHaveHeard(atMs: UInt64) -> Bool { gate.couldHaveHeard(atMs: atMs) }

    /// Makes the microphone and the speaker what the gate says, and publishes the one state.
    /// Called with ``lock`` held after every change.
    private func apply(_ now: UInt64) {
        microphone.isEnabled = gate.captureEnabled(nowMs: now)
        let shown = gate.displayed(nowMs: now)
        let speaker = !isPlaybackMuted && shown != .interrupted
        for receiver in connection.receivers {
            (receiver.track as? RTCAudioTrack)?.isEnabled = speaker
        }
        AudioSession.shared.publish(shown)
    }

    /// Makes this call's SDP offer.
    ///
    /// Section 15 ¶3: the client creates the offer. The host forwards it and never generates one.
    public func offer() async throws -> String {
        let constraints = RTCMediaConstraints(
            mandatoryConstraints: [
                kRTCMediaConstraintsOfferToReceiveAudio: kRTCMediaConstraintsValueTrue,
                kRTCMediaConstraintsOfferToReceiveVideo: kRTCMediaConstraintsValueFalse,
            ],
            optionalConstraints: nil
        )
        let description = try await connection.offer(for: constraints)
        try await connection.setLocalDescription(description)
        return description.sdp
    }

    /// Applies the provider's SDP answer.
    public func accept(answerSdp: String) async throws {
        let answer = RTCSessionDescription(type: .answer, sdp: answerSdp)
        do {
            try await connection.setRemoteDescription(answer)
        } catch {
            throw VoiceAudioError.answerNotApplicable(error.localizedDescription)
        }
    }

    /// Stops the person's voice reaching the model, without ending the call.
    ///
    /// Local and immediate. Section 15 ¶10 requires local microphone and speaker mute to remain
    /// available if the broker fails, so this touches the track and the audio session and nothing
    /// that could be waiting on a network answer.
    public func setMutedByPerson(_ muted: Bool) {
        lock.lock()
        defer { lock.unlock() }
        let now = VoiceCall.nowMs()
        gate.setMutedByPerson(muted, nowMs: now)
        apply(now)
    }

    /// Stops the model's voice coming out of this device, without ending the call.
    ///
    /// This is **playback**, and §15 ¶13 is explicit that speech interruption stops playback and
    /// not a coding task. Nothing here cancels anything on a host.
    public func setPlaybackMuted(_ muted: Bool) {
        lock.lock()
        defer { lock.unlock() }
        isPlaybackMuted = muted
        apply(VoiceCall.nowMs())
    }

    /// Ends the call and gives the microphone back.
    ///
    /// Local and immediate, for the same reason mute is.
    public func stop() {
        lock.lock()
        defer { lock.unlock() }
        guard !stopped else { return }
        stopped = true
        gate.stop(nowMs: VoiceCall.nowMs())
        expiry?.cancel()
        expiry = nil
        microphone.isEnabled = false
        connection.close()
        try? AudioSession.shared.deactivate()
    }
}

extension VoiceCall: AudioSessionEvents {
    public func audioSessionInterruption(began: Bool, mayResume: Bool) {
        lock.lock()
        defer { lock.unlock() }
        guard !stopped else { return }
        let now = VoiceCall.nowMs()
        // Began, or ended without the system's word to resume: the microphone stays closed. What
        // the person chose, their own mute, is untouched either way and applies when it reopens.
        let taken: VoiceCaptureGate.Taken = began ? .interrupted : (mayResume ? .none : .suspended)
        gate.taken(taken, nowMs: now)
        apply(now)
    }

    public func audioSessionRoute(changing: Bool, hasInput: Bool) {
        lock.lock()
        defer { lock.unlock() }
        guard !stopped else { return }
        let now = VoiceCall.nowMs()
        gate.route(changing: changing, inputAvailable: hasInput, nowMs: now)
        apply(now)
    }

    public func audioSessionReset() {
        lock.lock()
        defer { lock.unlock() }
        guard !stopped else { return }
        let now = VoiceCall.nowMs()
        gate.taken(.suspended, nowMs: now)
        apply(now)
    }
}

extension VoiceCall: RTCPeerConnectionDelegate {
    public func peerConnection(
        _: RTCPeerConnection,
        didChange state: RTCPeerConnectionState
    ) {
        observer?.voiceCall(self, connectionChanged: state)
    }

    public func peerConnection(_: RTCPeerConnection, didAdd receiver: RTCRtpReceiver, streams _: [RTCMediaStream]) {
        guard receiver.track is RTCAudioTrack, !announcedFirstAudio else { return }
        announcedFirstAudio = true
        observer?.voiceCallReceivedFirstAudio(self)
    }

    public func peerConnection(_: RTCPeerConnection, didOpen channel: RTCDataChannel) {
        // The provider opened its channel. This end reads it and never writes to it.
        channel.delegate = self
    }

    public func peerConnectionShouldNegotiate(_: RTCPeerConnection) {}
    public func peerConnection(_: RTCPeerConnection, didChange _: RTCSignalingState) {}
    public func peerConnection(_: RTCPeerConnection, didAdd _: RTCMediaStream) {}
    public func peerConnection(_: RTCPeerConnection, didRemove _: RTCMediaStream) {}
    public func peerConnection(_: RTCPeerConnection, didChange _: RTCIceConnectionState) {}
    public func peerConnection(_: RTCPeerConnection, didChange _: RTCIceGatheringState) {}
    public func peerConnection(_: RTCPeerConnection, didGenerate _: RTCIceCandidate) {}
    public func peerConnection(_: RTCPeerConnection, didRemove _: [RTCIceCandidate]) {}
}

extension VoiceCall: RTCDataChannelDelegate {
    public func dataChannelDidChangeState(_: RTCDataChannel) {}

    public func dataChannel(_: RTCDataChannel, didReceiveMessageWith buffer: RTCDataBuffer) {
        // Handed up as bytes. What is a known event and what is not is decided by the frozen
        // provider profile, one level above this file, and an unknown one is dropped there.
        observer?.voiceCall(self, receivedProviderEvent: buffer.data)
    }
}

File: apps/companion/native/android/android/src/main/java/to/kala/reach/companion/voice/VoiceCall.kt
package to.kala.reach.companion.voice

import android.content.Context
import android.media.AudioDeviceCallback
import android.media.AudioDeviceInfo
import android.media.AudioManager
import android.os.Handler
import android.os.HandlerThread
import android.os.SystemClock
import android.telephony.TelephonyManager
import org.webrtc.AudioTrack
import org.webrtc.DataChannel
import org.webrtc.DefaultVideoDecoderFactory
import org.webrtc.DefaultVideoEncoderFactory
import org.webrtc.EglBase
import org.webrtc.MediaConstraints
import org.webrtc.MediaStream
import org.webrtc.PeerConnection
import org.webrtc.PeerConnectionFactory
import org.webrtc.RtpReceiver
import org.webrtc.SdpObserver
import org.webrtc.SessionDescription
import org.webrtc.audio.JavaAudioDeviceModule
import to.kala.reach.companion.mobile.VoiceCallControl
import to.kala.reach.companion.mobile.VoiceCallPlatform
import to.kala.reach.companion.mobile.VoiceCaptureState
import to.kala.reach.companion.mobile.VoiceMediaSwitches
import to.kala.reach.companion.push.AudioSession
import java.nio.ByteBuffer
import java.util.concurrent.CountDownLatch
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong

/**
 * The native voice call: WebRTC's own peer connection, and the microphone the platform owns.
 *
 * Section 15 paragraph 2 is the constraint this file exists to satisfy, and it admits no exception:
 * native WebRTC and native platform audio own capture and playback, not a background WebView
 * `getUserMedia` path. Nothing here goes near the WebView. The offer is made by this process, the
 * answer is applied by this process, and the audio flows between this device and the provider
 * without passing through the interface or through KalaReach.
 *
 * Section 15 paragraph 6 is the other constraint: the provider's data channel is read-only. This
 * connection creates no channel of its own and sends zero bytes on the one the provider opens.
 *
 * Every decision about the microphone is [VoiceCallControl]'s; this class carries them out. The
 * platform's recorder is held off from the moment the connection exists, before anything is
 * negotiated, so applying an answer starts nothing on its own. The control turns it on only for a
 * call the host permitted whose foreground service holds the foreground, and every frame the
 * recorder produces is asked about separately and silenced unless the control says it may be
 * carried. Timers and the platform's reports run on this call's own thread, never the main one.
 */
class VoiceCall private constructor(
    private val context: Context,
    private val observer: Observer,
    /** This call's identity in the process, which the service's actions name. */
    val id: Long,
) {
    /** What a running call tells the application about. */
    interface Observer {
        /** The microphone's state changed, for the screen and for the authority gate. */
        fun onCaptureState(state: VoiceCaptureState)

        /** The provider sent something on its read-only channel. */
        fun onProviderEvent(bytes: ByteArray)

        /** The first remote audio arrived. KR-PERF-010's first-audio figure is taken here. */
        fun onFirstAudio()
    }

    /** The call's own thread: its timers, and everything the platform reports. */
    private val thread = HandlerThread("kr-voice-call-$id").apply { start() }
    private val handler = Handler(thread.looper)

    /** Held by every change to the connection and the tracks; never while waiting on WebRTC. */
    private val lock = Any()
    private var connection: PeerConnection? = null
    private var microphone: AudioTrack? = null
    private var announcedFirstAudio = false
    private var routeCallback: AudioDeviceCallback? = null
    private var audioDeviceOn: Boolean? = null

    /**
     * The one focus request this call holds. Its listener reports on this call's thread.
     *
     * The instance that asked for focus is the instance that gives it back.
     */
    private val audioSession =
        AudioSession(context) { change -> handler.post { onFocusChange(change) } }

    /**
     * The platform audio device, built for this call.
     *
     * Its recorder reports starting, stopping and failing to the control, on this call's thread,
     * and it hands every frame to [VoiceCallControl.carries] before the frame goes anywhere: a frame
     * the control does not let through is replaced with silence where it lies.
     */
    private val audioDevice: JavaAudioDeviceModule =
        JavaAudioDeviceModule.builder(context)
            .setAudioRecordStateCallback(
                object : JavaAudioDeviceModule.AudioRecordStateCallback {
                    override fun onWebRtcAudioRecordStart() {
                        handler.post { control.recorder(true) }
                    }

                    override fun onWebRtcAudioRecordStop() {
                        handler.post { control.recorder(false) }
                    }
                },
            )
            .setAudioRecordErrorCallback(
                object : JavaAudioDeviceModule.AudioRecordErrorCallback {
                    override fun onWebRtcAudioRecordInitError(message: String) {
                        handler.post { control.recorder(false) }
                    }

                    override fun onWebRtcAudioRecordStartError(
                        code: JavaAudioDeviceModule.AudioRecordStartErrorCode,
                        message: String,
                    ) {
                        handler.post { control.recorder(false) }
                    }

                    override fun onWebRtcAudioRecordError(message: String) {
                        handler.post { control.recorder(false) }
                    }
                },
            )
            .setAudioBufferCallback { buffer, _, _, _, bytesRead, captureTimeNs ->
                // On the audio thread, for every frame, before it reaches the encoder. Answered
                // from the gate alone, so no switch is set from here.
                if (!control.carries()) {
                    for (index in 0 until minOf(bytesRead, buffer.capacity())) buffer.put(index, 0)
                }
                captureTimeNs
            }
            .createAudioDeviceModule()
            .also { it.setMicrophoneMute(true) }

    private val egl: EglBase = EglBase.create()
    private val factory: PeerConnectionFactory =
        PeerConnectionFactory.builder()
            // Audio processing is the platform's where the device offers it: the audio device
            // uses the built-in acoustic echo canceller and noise suppressor when they exist, and
            // its own otherwise. Either way it is one canceller, not two fighting.
            .setAudioDeviceModule(audioDevice)
            .setVideoEncoderFactory(DefaultVideoEncoderFactory(egl.eglBaseContext, true, true))
            .setVideoDecoderFactory(DefaultVideoDecoderFactory(egl.eglBaseContext))
            .createPeerConnectionFactory()

    /** Every decision about the microphone, the speaker and the end of this call. */
    private val control = VoiceCallControl(Platform(), Switches())

    /** Whether the person has muted their own microphone. */
    val isMutedByPerson: Boolean
        get() = control.isMutedByPerson

    /** Whether the person has silenced the model's voice on this device. */
    val isPlaybackMuted: Boolean
        get() = control.isPlaybackMuted

    companion object {
        /** How long a description may take to be created or applied before the call is closed. */
        private const val DESCRIPTION_TIMEOUT_SECONDS = 30L

        private val calls = AtomicLong()

        /**
         * Builds a call for its offer and its answer, with the recorder off.
         *
         * The call is claimed as this process's one call before anything is opened, in one step,
         * so two starts cannot both pass a check and both publish. Nothing here opens the
         * microphone: that waits for [permit], the foreground service and the recorder.
         */
        @JvmStatic
        fun start(context: Context, observer: Observer): VoiceCall {
            PeerConnectionFactory.initialize(
                PeerConnectionFactory.InitializationOptions.builder(context.applicationContext)
                    .createInitializationOptions(),
            )
            val call = VoiceCall(context.applicationContext, observer, calls.incrementAndGet())
            // One microphone, one call. A second call started over a running one would leave the
            // first one's track and connection live with nothing owning them.
            if (!VoiceCallHolder.claim(call)) {
                call.stop()
                throw IllegalStateException("a voice call is already running on this device")
            }
            try {
                call.open()
            } catch (failure: Throwable) {
                call.stop()
                throw failure
            }
            return call
        }
    }

    private fun open() {
        synchronized(lock) {
            if (control.isStopped) throw IllegalStateException("this call was stopped while it was opening")
            val configuration = PeerConnection.RTCConfiguration(emptyList()).apply {
                // The provider's answer names its own candidates. No KalaReach relay is
                // configured: media travels between this device and the provider, and a relay of
                // KalaReach's would be a third party in a path section 15 paragraph 3 says has two.
                sdpSemantics = PeerConnection.SdpSemantics.UNIFIED_PLAN
                continualGatheringPolicy =
                    PeerConnection.ContinualGatheringPolicy.GATHER_CONTINUALLY
            }
            val made = factory.createPeerConnection(configuration, PeerObserver())
                ?: throw IllegalStateException("this device could not open a connection")
            // Off before anything is negotiated. Left on, applying the answer would start the
            // recorder by itself, whether or not the host ever permitted the call.
            made.setAudioRecording(false)
            made.setAudioPlayout(false)
            audioDeviceOn = false
            connection = made
            val source = factory.createAudioSource(MediaConstraints())
            // Off until the control turns it on. The offer describes a track, and a described
            // track carries nothing until the control lets it.
            microphone = factory.createAudioTrack("kr-voice-microphone", source).also {
                it.setEnabled(false)
                made.addTrack(it, listOf("kr-voice"))
            }
        }
        watchRoute()
    }

    /**
     * Opens the microphone for a call the host started.
     *
     * Takes the host's answer: the voice session and the moment the service closes the call, in
     * UTC milliseconds. Focus and the foreground service are asked for here; the recorder waits
     * for the service to hold the foreground and then for the recorder itself to report running.
     *
     * @return false, and the call ended, when the moment has passed or the platform refused focus
     * or the service; false and nothing changed when this call is stopped or already permitted.
     */
    fun permit(voiceSessionId: String, closesAtEpochMs: Long): Boolean =
        control.permit(voiceSessionId, closesAtEpochMs)

    /** The foreground service holds the foreground. Called by the service, for this call only. */
    fun servicePromoted() {
        handler.post { control.servicePromoted() }
    }

    /** The foreground service could not enter the foreground. The call ends. */
    fun serviceRefused() {
        handler.post { control.serviceRefused() }
    }

    /**
     * Makes this call's SDP offer.
     *
     * Section 15 paragraph 3: the client creates the offer. The host forwards it and never
     * generates one.
     */
    fun offer(): String {
        val constraints = MediaConstraints().apply {
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveAudio", "true"))
            mandatory.add(MediaConstraints.KeyValuePair("OfferToReceiveVideo", "false"))
        }
        val open = openConnection()
        val made = awaitDescription { observer -> open.createOffer(observer, constraints) }
        awaitSet { observer -> open.setLocalDescription(observer, made) }
        return made.description
    }

    /** Applies the provider's SDP answer. It starts nothing: the recorder stays off until [permit]. */
    fun accept(answerSdp: String) {
        val answer = SessionDescription(SessionDescription.Type.ANSWER, answerSdp)
        val open = openConnection()
        awaitSet { observer -> open.setRemoteDescription(observer, answer) }
    }

    /** The connection, when the call is still open. */
    private fun openConnection(): PeerConnection =
        synchronized(lock) {
            connection.takeIf { !control.isStopped }
                ?: throw IllegalStateException("this voice call has ended")
        }

    /**
     * Stops the person's voice reaching the model, without ending the call.
     *
     * Local and immediate. Section 15 paragraph 10 requires local microphone and speaker mute to
     * remain available if the broker fails, so this touches the track and nothing that could be
     * waiting on a network answer. Unmuting is the person's choice and nothing more: it does not
     * open a microphone the system has taken or a call that has not been permitted.
     */
    fun setMutedByPerson(muted: Boolean) = control.setMutedByPerson(muted)

    /**
     * Stops the model's voice coming out of this device, without ending the call.
     *
     * This is playback, and section 15 paragraph 13 is explicit that speech interruption stops
     * playback and not a coding task. Nothing here cancels anything on a host.
     */
    fun setPlaybackMuted(muted: Boolean) = control.setPlaybackMuted(muted)

    /**
     * Whether the microphone was carrying speech at `atMs` on the monotonic clock.
     *
     * What a claim that something was said is checked against, from this device's own record.
     */
    fun couldHaveHeard(atMs: Long): Boolean = control.couldHaveHeard(atMs)

    /** Ends the call and gives the microphone back. Local and immediate, as mute is. */
    fun stop() = control.stop()

    /** Acts on what the system did to audio focus. */
    private fun onFocusChange(change: Int) {
        when (change) {
            AudioManager.AUDIOFOCUS_LOSS, AudioManager.AUDIOFOCUS_LOSS_TRANSIENT ->
                control.focus(lost = true)
            AudioManager.AUDIOFOCUS_GAIN -> control.focus(lost = false)
            else -> Unit
        }
    }

    /** True while the platform reports a call on the cellular radio. */
    @Suppress("DEPRECATION")
    fun isInPhoneCall(): Boolean =
        context.getSystemService(TelephonyManager::class.java)?.callState !=
            TelephonyManager.CALL_STATE_IDLE

    private fun hasInput(manager: AudioManager): Boolean =
        manager.getDevices(AudioManager.GET_DEVICES_INPUTS).isNotEmpty()

    private fun watchRoute() {
        val manager = context.getSystemService(AudioManager::class.java) ?: return
        val callback = object : AudioDeviceCallback() {
            override fun onAudioDevicesAdded(added: Array<out AudioDeviceInfo>?) = changed()

            override fun onAudioDevicesRemoved(removed: Array<out AudioDeviceInfo>?) = changed()

            private fun changed() {
                // A Bluetooth headset arriving or leaving. The state says so while capture is
                // re-established rather than claiming the person is still being heard.
                control.route(changing = true, inputAvailable = true)
                control.route(changing = false, inputAvailable = hasInput(manager))
            }
        }
        synchronized(lock) { routeCallback = callback }
        // Reported on this call's own thread.
        manager.registerAudioDeviceCallback(callback, handler)
        handler.post { control.route(changing = false, inputAvailable = hasInput(manager)) }
    }

    /** The platform, as the control sees it. */
    private inner class Platform : VoiceCallPlatform {
        override fun nowMs(): Long = SystemClock.elapsedRealtime()

        override fun epochMs(): Long = System.currentTimeMillis()

        override fun acquireFocus(): Boolean = audioSession.activate()

        override fun releaseFocus() = audioSession.deactivate()

        override fun startService(): Boolean {
            // The service is what keeps capture alive once the screen locks, and it is asked for
            // for a call the host permitted rather than by anything that could fire on its own.
            VoiceCallService.start(context, id)
            return true
        }

        override fun stopService() {
            try {
                VoiceCallService.stop(context, id)
            } catch (notRunning: IllegalStateException) {
                // The platform refuses to reach a service it would have to start from the
                // background, which means none is running for this call: nothing to stop.
            }
        }

        override fun schedule(atMs: Long, task: () -> Unit): () -> Unit {
            val runnable = Runnable { task() }
            handler.postDelayed(runnable, maxOf(0L, atMs - SystemClock.elapsedRealtime()))
            return { handler.removeCallbacks(runnable) }
        }

        override fun publish(state: VoiceCaptureState) = observer.onCaptureState(state)

        override fun publishToService(state: VoiceCaptureState) {
            try {
                VoiceCallService.publishCapture(context, id, state)
            } catch (notRunning: IllegalStateException) {
                // No service is running for this call, so there is no notification to update.
            }
        }

        override fun ended() {
            val (open, callback) = synchronized(lock) {
                val held = connection to routeCallback
                connection = null
                microphone = null
                routeCallback = null
                held
            }
            // Cleared only if this call is the one published: a call that was already replaced
            // must not take its replacement's place in the holder with it.
            VoiceCallHolder.claimStopped(this@VoiceCall)
            callback?.let {
                context.getSystemService(AudioManager::class.java)?.unregisterAudioDeviceCallback(it)
            }
            open?.dispose()
            factory.dispose()
            audioDevice.release()
            egl.release()
            thread.quitSafely()
        }
    }

    /** The media, as the control sets it. */
    private inner class Switches : VoiceMediaSwitches {
        override fun setAudioDevice(on: Boolean) {
            val open = synchronized(lock) {
                if (audioDeviceOn == on) return
                audioDeviceOn = on
                connection
            } ?: return
            open.setAudioRecording(on)
            open.setAudioPlayout(on)
        }

        override fun setMicrophone(on: Boolean) {
            audioDevice.setMicrophoneMute(!on)
            synchronized(lock) { microphone }?.setEnabled(on)
        }

        override fun setPlayback(on: Boolean) {
            val receivers = synchronized(lock) { connection }?.receivers ?: return
            receivers.forEach { (it.track() as? AudioTrack)?.setEnabled(on) }
        }
    }

    private inner class PeerObserver : PeerConnection.Observer {
        override fun onDataChannel(channel: DataChannel) {
            // The provider opened its channel. This end reads it and never writes to it.
            channel.registerObserver(object : DataChannel.Observer {
                override fun onBufferedAmountChange(previous: Long) {}
                override fun onStateChange() {}
                override fun onMessage(buffer: DataChannel.Buffer) {
                    val bytes = ByteArray(buffer.data.remaining())
                    (buffer.data as ByteBuffer).get(bytes)
                    // Handed up as bytes. What is a known event is decided by the frozen provider
                    // profile, one level above this file, and an unknown one is dropped there.
                    observer.onProviderEvent(bytes)
                }
            })
        }

        override fun onAddTrack(receiver: RtpReceiver, streams: Array<out MediaStream>) {
            if (receiver.track() !is AudioTrack) return
            // A track arrives playing; the control decides whether it may.
            handler.post { control.refresh() }
            if (announcedFirstAudio) return
            announcedFirstAudio = true
            observer.onFirstAudio()
        }

        override fun onSignalingChange(state: PeerConnection.SignalingState) {}
        override fun onIceConnectionChange(state: PeerConnection.IceConnectionState) {}
        override fun onIceConnectionReceivingChange(receiving: Boolean) {}
        override fun onIceGatheringChange(state: PeerConnection.IceGatheringState) {}
        override fun onIceCandidate(candidate: org.webrtc.IceCandidate) {}
        override fun onIceCandidatesRemoved(candidates: Array<out org.webrtc.IceCandidate>) {}
        override fun onAddStream(stream: MediaStream) {}
        override fun onRemoveStream(stream: MediaStream) {}
        override fun onRenegotiationNeeded() {}
    }

    private fun awaitDescription(ask: (SdpObserver) -> Unit): SessionDescription {
        val latch = CountDownLatch(1)
        var made: SessionDescription? = null
        var failure: String? = null
        ask(object : SimpleSdpObserver() {
            override fun onCreateSuccess(description: SessionDescription) {
                made = description
                latch.countDown()
            }

            override fun onCreateFailure(reason: String) {
                failure = reason
                latch.countDown()
            }
        })
        if (!latch.await(DESCRIPTION_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
            stop()
            throw IllegalStateException("this device did not finish making an offer in time")
        }
        return made ?: throw IllegalStateException(failure ?: "this device could not make an offer")
    }

    /**
     * Waits for one description to be applied.
     *
     * A timeout is a failure, not a success. Returning normally when no callback arrived would
     * leave a half-negotiated connection that the caller believes is ready, so the call is closed
     * and the caller is told.
     */
    private fun awaitSet(ask: (SdpObserver) -> Unit) {
        val latch = CountDownLatch(1)
        var failure: String? = null
        ask(object : SimpleSdpObserver() {
            override fun onSetSuccess() = latch.countDown()

            override fun onSetFailure(reason: String) {
                failure = reason
                latch.countDown()
            }
        })
        if (!latch.await(DESCRIPTION_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
            stop()
            throw IllegalStateException("this device did not apply a session description in time")
        }
        failure?.let {
            stop()
            throw IllegalStateException(it)
        }
    }

    private abstract class SimpleSdpObserver : SdpObserver {
        override fun onCreateSuccess(description: SessionDescription) {}
        override fun onSetSuccess() {}
        override fun onCreateFailure(reason: String) {}
        override fun onSetFailure(reason: String) {}
    }
}

/**
 * The one call this process has, so the notification's actions reach it.
 *
 * A running call is process-wide because the microphone is, and the notification's mute and stop
 * are pressed while no interface is running at all. Every action names the call it was for, so an
 * action that arrives late reaches that call or nothing, never the call that replaced it.
 */
object VoiceCallHolder {
    @JvmStatic
    @Volatile
    var current: VoiceCall? = null
        private set

    /** Publishes `call` as this process's call, in one step, unless one is already running. */
    @JvmStatic
    @Synchronized
    fun claim(call: VoiceCall): Boolean {
        if (current != null) return false
        current = call
        return true
    }

    /** The running call, when it is the one `id` names. */
    @JvmStatic
    fun current(id: Long): VoiceCall? = current?.takeIf { it.id == id }

    /**
     * Clears the holder, but only when the call that stopped is the one it holds.
     *
     * A call that was already replaced clearing the holder would leave its replacement running with
     * nothing published, and the service's stop would then find nothing to stop.
     */
    @JvmStatic
    @Synchronized
    fun claimStopped(call: VoiceCall) {
        if (current === call) current = null
    }
}
